odl/test/discr/discr_utils_test.py

```diff
@@ -0,0 +1,265 @@
+    f = np.arange(1, 9, dtype='float64').reshape((4, 2))
+    interpolator = linear_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.6])
+    l1 = (0.3 - 0.125) / (0.375 - 0.125)
+    l2 = (0.6 - 0.25) / (0.75 - 0.25)
+    true_val = (
+        (1 - l1) * (1 - l2) * f[0, 0]
+        + (1 - l1) * l2 * f[0, 1]
+        + l1 * (1 - l2) * f[1, 0]
+        + l1 * l2 * f[1, 1]
+    )
+    assert val == pytest.approx(true_val)
+
+    # Input array, with and without output array
+    pts = np.array([[0.3, 0.6],
+                    [0.1, 0.25],
+                    [1.0, 1.0]])
+    l1 = (0.3 - 0.125) / (0.375 - 0.125)
+    l2 = (0.6 - 0.25) / (0.75 - 0.25)
+    true_val_1 = (
+        (1 - l1) * (1 - l2) * f[0, 0]
+        + (1 - l1) * l2 * f[0, 1]
+        + l1 * (1 - l2) * f[1, 0]
+        + l1 * l2 * f[1, 1]
+    )
+    l1 = (0.125 - 0.1) / (0.375 - 0.125)
+    # l2 = 0
+    true_val_2 = (1 - l1) * f[0, 0]  # only lower left contributes
+    l1 = (1.0 - 0.875) / (0.875 - 0.625)
+    l2 = (1.0 - 0.75) / (0.75 - 0.25)
+    true_val_3 = (1 - l1) * (1 - l2) * f[3, 1]  # lower left only
+    true_arr = [true_val_1, true_val_2, true_val_3]
+    assert all_equal(interpolator(pts.T), true_arr)
+
+    out = np.empty(3, dtype='float64')
+    interpolator(pts.T, out=out)
+    assert all_equal(out, true_arr)
+
+    # Input meshgrid, with and without output array
+    mg = sparse_meshgrid([0.3, 1.0], [0.4, 0.75])
+    # Indices: (1, 3) x (0, 1)
+    lx1 = (0.3 - 0.125) / (0.375 - 0.125)
+    lx2 = (1.0 - 0.875) / (0.875 - 0.625)
+    ly1 = (0.4 - 0.25) / (0.75 - 0.25)
+    # ly2 = 0
+    true_val_11 = (
+        (1 - lx1) * (1 - ly1) * f[0, 0]
+        + (1 - lx1) * ly1 * f[0, 1]
+        + lx1 * (1 - ly1) * f[1, 0]
+        + lx1 * ly1 * f[1, 1]
+    )
+    true_val_12 = (
+        (1 - lx1) * f[0, 1]
+        + lx1 * f[1, 1]  # ly2 = 0
+    )
+    true_val_21 = (
+        (1 - lx2) * (1 - ly1) * f[3, 0]
+        + (1 - lx2) * ly1 * f[3, 1]   # high node 1.0, no upper
+    )
+    true_val_22 = (1 - lx2) * f[3, 1]  # ly2 = 0, no upper for 1.0
+    true_mg = [[true_val_11, true_val_12],
+               [true_val_21, true_val_22]]
+    assert all_equal(interpolator(mg), true_mg)
+    out = np.empty((2, 2), dtype='float64')
+    interpolator(mg, out=out)
+    assert all_equal(out, true_mg)
+
+
+def test_per_axis_interpolation():
+    """Test different interpolation schemes per axis."""
+    coord_vecs = [[0.125, 0.375, 0.625, 0.875], [0.25, 0.75]]
+    interp = ['linear', 'nearest']
+    f = np.arange(1, 9, dtype='float64').reshape((4, 2))
+    interpolator = per_axis_interpolator(f, coord_vecs, interp)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.5])
+    l1 = (0.3 - 0.125) / (0.375 - 0.125)
+    # 0.5 equally far from both neighbors -> NN chooses 0.75
+    true_val = (1 - l1) * f[0, 1] + l1 * f[1, 1]
+    assert val == pytest.approx(true_val)
```

Comment:
Again, I only did this one carefully. But the rest have the same structure and therefore seems correct too.